# Test RAG Pipeline With LangChain

In [6]:
#!pip freeze | grep langchain

In [2]:
import sys
from pathlib import Path


# Detect project root correctly:
# Notebook CWD = /project-root/notebooks → parent = /project-root
project_root = Path.cwd().parent
print("Detected project root:", project_root)

# Add project root to Python import path (so vector_pipeline can be imported cleanly)
if str(project_root) not in sys.path:
    sys.path.insert(0, str(project_root))
    print("Added to sys.path.")

Detected project root: /Users/muhammetsantas/Downloads/EDUCATION/POWERCODERS/PERSONAL_PROJECTS/chatbot/project-25-2-scrum-team-data
Added to sys.path.


In [3]:
%%time
from rag_pipeline.rag_pipeline import ask_question

query = "I want to have a YAMAHA electric guitar which is the most rated product in the site."

response = ask_question(question=query, k=10, use_reranker=False)

print("Question:", query)
print("\nAnswer:\n", response)

/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


Loading existing Chroma DB from /Users/muhammetsantas/Downloads/EDUCATION/POWERCODERS/PERSONAL_PROJECTS/chatbot/project-25-2-scrum-team-data/chroma_db ...


/Users/muhammetsantas/Downloads/EDUCATION/POWERCODERS/PERSONAL_PROJECTS/chatbot/project-25-2-scrum-team-data/rag_pipeline/rag_pipeline.py:214: LangChainDeprecationWarning: The class `Chroma` was deprecated in LangChain 0.2.9 and will be removed in 1.0. An updated version of the class exists in the `langchain-chroma package and should be used instead. To use it run `pip install -U `langchain-chroma` and import as `from `langchain_chroma import Chroma``.
  vectorstore = Chroma(


Vectorstore loaded successfully.
Document count: 6401
Prompt was created successfully.
✅ API key was loaded successfully!
📍 .env file location: /Users/muhammetsantas/Downloads/EDUCATION/POWERCODERS/PERSONAL_PROJECTS/chatbot/project-25-2-scrum-team-data/.env
LLM instance created successfully.
Question: I want to have a YAMAHA electric guitar which is the most rated product in the site.

Answer:
 <p>The most highly-rated YAMAHA electric guitar is the Yamaha Pacifica Series PAC012DLX Electric Guitar, with a rating of 4.6 out of 5 stars based on 671 ratings. It is priced at $209.99 and comes in Old Violin Sunburst.</p>
<p><strong>Key Features</strong></p>
<ul>
<li>Agathis Body</li>
<li>Maple Bolt-On Neck</li>
<li>Sonokeling Fingerboard</li>
<li>Vintage Tremelo</li>
<li>5 Position Switch</li>
</ul>
<p><strong>Specifications / Metadata</strong></p>
<ul>
<li>Color: Old Violin Sunburst</li>
<li>Price: $209.99</li>
<li>Rating: 4.6</li>
<li>Rating Count: 671</li>
<li>Store: YAMAHA</li>
</ul>
CPU

In [3]:
response = ask_question(question=query, k=10, use_reranker=False)

print("Question:", query)
print("\nAnswer:\n", response)

Prompt was created successfully.
✅ API key was loaded successfully!
📍 .env file location: /Users/muhammetsantas/Downloads/EDUCATION/POWERCODERS/PERSONAL_PROJECTS/chatbot/project-25-2-scrum-team-data/.env
LLM instance created successfully.
Question: I want to have a YAMAHA electric guitar which is the most rated product in the site.

Answer:
 <p>The most highly rated Yamaha electric guitar in the provided information is the Yamaha Pacifica Series PAC012DLX.</p>
<p><strong>Key Features</strong></p>
<ul>
<li>Agathis Body</li>
<li>Maple Bolt-On Neck</li>
<li>Sonokeling Fingerboard</li>
<li>Vintage Tremelo</li>
<li>5 Position Switch</li>
</ul>
<p><strong>Specifications / Metadata</strong></p>
<ul>
<li><strong>Store:</strong> YAMAHA</li>
<li><strong>Color:</strong> Old Violin Sunburst</li>
<li><strong>Rating:</strong> 4.6</li>
<li><strong>Rating Count:</strong> 671</li>
<li><strong>Price:</strong> 209.99</li>
<li><strong>Product ID:</strong> B0757Q45NK</li>
</ul>
<p><strong>Summary</strong>


In [4]:
import json

def print_json(data):
    print(json.dumps(data, indent=2, ensure_ascii=False))

In [5]:
query = "I want to have a YAMAHA electric guitar which is the most rated product in the site."

api_request = {
    "session_id": "S12345",
    "user_id": "U9001",
    "content": query,
    "top_k": 10,
    "use_reranker": False
}

print("📥 Incoming API request:")
print_json(api_request)

📥 Incoming API request:
{
  "session_id": "S12345",
  "user_id": "U9001",
  "content": "I want to have a YAMAHA electric guitar which is the most rated product in the site.",
  "top_k": 10,
  "use_reranker": false
}


In [6]:
result = ask_question(
    question = api_request["content"],
    k=api_request.get("top_k", 5),
    use_reranker=api_request.get("use_reranker", False)
)

api_response = {
    "status": "success",
    "session_id": api_request["session_id"],
    "user_id": api_request["user_id"],
    "answer": result
}

print("📤 API response:")
print_json(api_response)

Prompt was created successfully.
✅ API key was loaded successfully!
📍 .env file location: /Users/muhammetsantas/Downloads/EDUCATION/POWERCODERS/PERSONAL_PROJECTS/chatbot/project-25-2-scrum-team-data/.env
LLM instance created successfully.
📤 API response:
{
  "status": "success",
  "session_id": "S12345",
  "user_id": "U9001",
  "answer": "<p>The most highly rated Yamaha electric guitar in the provided information is the Yamaha Pacifica Series PAC012DLX Electric Guitar in Old Violin Sunburst, with a rating of 4.6 based on 671 ratings.</p>\n<p><strong>Key Features</strong></p>\n<ul>\n<li>Agathis Body</li>\n<li>Maple Bolt-On Neck</li>\n<li>Sonokeling Fingerboard</li>\n<li>Vintage Tremelo</li>\n<li>5 Position Switch</li>\n</ul>\n<p><strong>Specifications / Metadata</strong></p>\n<ul>\n<li><strong>Store:</strong> YAMAHA</li>\n<li><strong>Color:</strong> Old Violin Sunburst</li>\n<li><strong>Rating:</strong> 4.6</li>\n<li><strong>Rating Count:</strong> 671</li>\n<li><strong>Price:</strong> 2

In [7]:
print(api_response["answer"])

<p>The most highly rated Yamaha electric guitar in the provided information is the Yamaha Pacifica Series PAC012DLX Electric Guitar in Old Violin Sunburst, with a rating of 4.6 based on 671 ratings.</p>
<p><strong>Key Features</strong></p>
<ul>
<li>Agathis Body</li>
<li>Maple Bolt-On Neck</li>
<li>Sonokeling Fingerboard</li>
<li>Vintage Tremelo</li>
<li>5 Position Switch</li>
</ul>
<p><strong>Specifications / Metadata</strong></p>
<ul>
<li><strong>Store:</strong> YAMAHA</li>
<li><strong>Color:</strong> Old Violin Sunburst</li>
<li><strong>Rating:</strong> 4.6</li>
<li><strong>Rating Count:</strong> 671</li>
<li><strong>Price:</strong> 209.99</li>
<li><strong>Product ID:</strong> B0757Q45NK</li>
</ul>
<p><strong>Summary</strong>
This is a solid body electric guitar with a vintage tremelo and 5-position switch, designed for a variety of playing styles. It is well-regarded by customers, as indicated by its high rating and number of reviews.</p>


In [ ]:
result = ask_question(
    question = api_request["content"],
    k=api_request.get("top_k", 5),
    use_reranker=api_request.get("use_reranker", True)
)

api_response = {
    "status": "success",
    "session_id": api_request["session_id"],
    "user_id": api_request["user_id"],
    "answer": result
}

print("📤 API response:")
print_json(api_response)

Prompt was created successfully.
✅ API key was loaded successfully!
📍 .env file location: /Users/muhammetsantas/Downloads/EDUCATION/POWERCODERS/PERSONAL_PROJECTS/chatbot/project-25-2-scrum-team-data/.env
LLM instance created successfully.
📤 API response:
{
  "status": "success",
  "session_id": "S12345",
  "user_id": "U9001",
  "answer": "<p>The most highly rated Yamaha electric guitar in the provided information is the Yamaha Pacifica Series PAC012DLX with a rating of 4.6 based on 671 ratings.</p>\n<p><strong>Key Features</strong></p>\n<ul>\n<li>Agathis Body</li>\n<li>Maple Bolt-On Neck</li>\n<li>Sonokeling Fingerboard</li>\n<li>Vintage Tremelo</li>\n<li>5 Position Switch</li>\n</ul>\n<p><strong>Specifications / Metadata</strong></p>\n<ul>\n<li><strong>Store:</strong> YAMAHA</li>\n<li><strong>Color:</strong> Old Violin Sunburst</li>\n<li><strong>Rating:</strong> 4.6</li>\n<li><strong>Rating Count:</strong> 671</li>\n<li><strong>Price:</strong> 209.99</li>\n<li><strong>Product ID:</str

In [ ]:
print(api_response["answer"])

<p>The most highly rated Yamaha electric guitar in the provided information is the Yamaha Pacifica Series PAC012DLX Electric Guitar, with a rating of 4.6 based on 671 ratings.</p>
<p><strong>Key Features</strong></p>
<ul>
<li>Agathis Body</li>
<li>Maple Bolt-On Neck</li>
<li>Sonokeling Fingerboard</li>
<li>Vintage Tremelo</li>
<li>5 Position Switch</li>
</ul>
<p><strong>Specifications / Metadata</strong></p>
<ul>
<li><strong>Store:</strong> YAMAHA</li>
<li><strong>Color:</strong> Old Violin Sunburst</li>
<li><strong>Rating:</strong> 4.6</li>
<li><strong>Rating Count:</strong> 671</li>
<li><strong>Price:</strong> 209.99</li>
<li><strong>Product ID:</strong> B0757Q45NK</li>
</ul>
<p><strong>Summary</strong>
This is a solid-body electric guitar from Yamaha's Pacifica series, known for its quality and playability. It features a vintage tremelo and 5-position switch for versatile sound options.</p>


## Multi Turn Chat

In [1]:
import sys
from pathlib import Path


# Detect project root correctly:
# Notebook CWD = /project-root/notebooks → parent = /project-root
project_root = Path.cwd().parent
print("Detected project root:", project_root)

# Add project root to Python import path (so vector_pipeline can be imported cleanly)
if str(project_root) not in sys.path:
    sys.path.insert(0, str(project_root))
    print("Added to sys.path.")

Detected project root: /Users/muhammetsantas/Downloads/EDUCATION/POWERCODERS/PERSONAL_PROJECTS/chatbot/project-25-2-scrum-team-data
Added to sys.path.


In [2]:
import sqlite3
import json
from pathlib import Path

project_root = Path.cwd().parent
db_path = project_root / "chat_history.db"

def print_chat_history_db():
    """Print chat sessions and html_answers stored in the DB (messages + html list)."""
    print(f"📁 Chat History DB Path: {db_path}")
    print(f"✅ Is there DB? {db_path.exists()}")

    sessions = {}

    if db_path.exists():
        conn = sqlite3.connect(str(db_path))
        cursor = conn.cursor()

        cursor.execute("SELECT name FROM sqlite_master WHERE type='table';")
        tables = cursor.fetchall()
        print(f"\n📊 Tables: {tables}")

        cursor.execute("SELECT * FROM chat_sessions;")
        rows = cursor.fetchall()

        if rows:
            print(f"\n📝 Total sessions: {len(rows)}")
            print("\n" + "="*80)
            for i, row in enumerate(rows, 1):
                print(f"\n🔹 Session {i}:")
                print(f"   ID: {row[0]}")
                sid = row[1] if len(row) > 1 else None
                print(f"   Session ID: {sid}")
                # Schema used by DB helper: (id, session_id, messages, html_answers, created_at, updated_at)
                created = row[4] if len(row) > 4 else None
                updated = row[5] if len(row) > 5 else None
                print(f"   Created At: {created}")
                print(f"   Updated At: {updated}")

                messages = json.loads(row[2]) if row[2] else []
                print(f"   Messages ({len(messages)} entries):")
                for j, msg in enumerate(messages, 1):
                    content = msg.get('content', '')
                    preview = content[:200] + "..." if len(content) > 200 else content
                    print(f"      {j}. Role: {msg.get('role')} | {preview}")

                # html_answers stored in column index 3 as JSON list
                html_list = []
                if len(row) > 3 and row[3]:
                    try:
                        html_list = json.loads(row[3])
                        if not isinstance(html_list, list):
                            html_list = [html_list]
                    except Exception:
                        html_list = [row[3]]

                if html_list:
                    print(f"\n   HTML Answers ({len(html_list)}):")
                    for k, html in enumerate(html_list, 1):
                        preview = html[:400] + "..." if len(html) > 400 else html
                        print(f"     {k}. {preview}")
                else:
                    print("   No HTML answers stored.")

                sessions[sid] = {"messages": messages, "html_answers": html_list, "created_at": created, "updated_at": updated}
        else:
            print("\n⚠️  DB empty (no chat yet).")

        conn.close()
    else:
        print("\n⚠️  DB file has not been created yet. Call ask_question() first.")

    return sessions


In [3]:
%%time
import random
from multi_turn_pipeline.rag_pipeline import ask_question

session_id = random.randint(100000, 999999)
print(f"session ID: {session_id}")

query = "I want to buy a cheap guitar for beginners. What are the most cheapest 3 options available?"

response = ask_question(question=query, k=10, use_reranker=False, session_id=session_id)

print("Question:", query)
print("\nAnswer:\n", response)

/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


session ID: 609387
✅ API key is present (loaded or already set).
LLM instance created and cached.
history: []
Loading existing Chroma DB from /Users/muhammetsantas/Downloads/EDUCATION/POWERCODERS/PERSONAL_PROJECTS/chatbot/project-25-2-scrum-team-data/chroma_db ...


/Users/muhammetsantas/Downloads/EDUCATION/POWERCODERS/PERSONAL_PROJECTS/chatbot/project-25-2-scrum-team-data/multi_turn_pipeline/rag_pipeline.py:215: LangChainDeprecationWarning: The class `Chroma` was deprecated in LangChain 0.2.9 and will be removed in 1.0. An updated version of the class exists in the `langchain-chroma package and should be used instead. To use it run `pip install -U `langchain-chroma` and import as `from `langchain_chroma import Chroma``.
  vectorstore = Chroma(


Vectorstore loaded successfully.
Document count: 6401
Prompt was created successfully.
Question: I want to buy a cheap guitar for beginners. What are the most cheapest 3 options available?

Answer:
 <p>Here are the three cheapest guitar options for beginners based on the provided information:</p>
<ol>
<li><strong>Smartxchoices starter kit:</strong> Priced at $30.99, this kit includes everything you need to get started, such as a guitar, strings, a pitch pipe, a shoulder strap, and a carrying case.</li>
<li><strong>Best Choice Products 38-inch Beginner Acoustic Guitar Package:</strong> This option costs $49.99 and comes with a gig bag, picks, a strap, a tuner, and an instructional sheet.</li>
<li><strong>Best Choice Products 30in Kids Acoustic Guitar Starter Kit:</strong> This kit is available for $44.99 and includes a carrying case, extra strings, a strap, a pick set, and a pitch pipe.</li>
</ol>
<p><strong>Summary:</strong> All three options are affordable and suitable for beginners, 

Question: I want to buy a cheap guitar for beginners. What are the most cheapest 3 options available?

Answer:
 <p>Here are the three cheapest guitar options for beginners based on the provided information:</p>
<ol>
<li><strong>Smartxchoices starter kit:</strong> Priced at $30.99, this kit includes everything you need to get started, such as a guitar, pitch pipe, guitar pick, shoulder strap, and an extra set of strings. It comes un-tuned and has steel strings.</li>
<li><strong>Best Choice Products 30in Kids Acoustic Guitar:</strong> This option is available for $44.99. It's designed for kids and includes a carrying case, extra strings, strap, pick set, and pitch pipe.</li>
<li><strong>Best Choice Products 38in Beginner Acoustic Guitar:</strong> This guitar starter kit is priced at $49.99 and includes an e-tuner, gig bag, guitar strap, extra nylon strings, picks, a capo, and a cloth.</li>
</ol>
<p>This dataset does not include real-time availability information.</p>

In [4]:
print_chat_history_db()

📁 Chat History DB Path: /Users/muhammetsantas/Downloads/EDUCATION/POWERCODERS/PERSONAL_PROJECTS/chatbot/project-25-2-scrum-team-data/chat_history.db
✅ Is there DB? True

📊 Tables: [('chat_sessions',), ('sqlite_sequence',)]

📝 Total sessions: 1


🔹 Session 1:
   ID: 1
   Session ID: 609387
   Created At: 2025-12-11 18:19:48
   Updated At: 2025-12-11 18:19:48
   Messages (2 entries):
      1. Role: user | I want to buy a cheap guitar for beginners. What are the most cheapest 3 options available?
      2. Role: assistant | Here are the three cheapest guitar options for beginners based on the provided information:

1.  **Smartxchoices starter kit:** Priced at $30.99, this kit includes everything you need to get started, ...

   HTML Answers (1):
     1. <p>Here are the three cheapest guitar options for beginners based on the provided information:</p>
<ol>
<li><strong>Smartxchoices starter kit:</strong> Priced at $30.99, this kit includes everything you need to get started, such as a guitar

{'609387': {'messages': [{'role': 'user',
    'content': 'I want to buy a cheap guitar for beginners. What are the most cheapest 3 options available?'},
   {'role': 'assistant',
    'content': 'Here are the three cheapest guitar options for beginners based on the provided information:\n\n1.  **Smartxchoices starter kit:** Priced at $30.99, this kit includes everything you need to get started, such as a guitar, strings, a pitch pipe, a shoulder strap, and a carrying case.\n2.  **Best Choice Products 38-inch Beginner Acoustic Guitar Package:** This option costs $49.99 and comes with a gig bag, picks, a strap, a tuner, and an instructional sheet.\n3.  **Best Choice Products 30in Kids Acoustic Guitar Starter Kit:** This kit is available for $44.99 and includes a carrying case, extra strings, a strap, a pick set, and a pitch pipe.\n\n**Summary:** All three options are affordable and suitable for beginners, with prices ranging from $30.99 to $49.99.\n\n**Metadata:**\n*   **Smartxchoices:** C

In [5]:
%%time
query = "Which one is the best option?"

response = ask_question(question=query, k=5, use_reranker=False, session_id=session_id)

print("Question:", query)
print("\nAnswer:\n", response)

history: [{'role': 'user', 'content': 'I want to buy a cheap guitar for beginners. What are the most cheapest 3 options available?'}, {'role': 'assistant', 'content': 'Here are the three cheapest guitar options for beginners based on the provided information:\n\n1.  **Smartxchoices starter kit:** Priced at $30.99, this kit includes everything you need to get started, such as a guitar, strings, a pitch pipe, a shoulder strap, and a carrying case.\n2.  **Best Choice Products 38-inch Beginner Acoustic Guitar Package:** This option costs $49.99 and comes with a gig bag, picks, a strap, a tuner, and an instructional sheet.\n3.  **Best Choice Products 30in Kids Acoustic Guitar Starter Kit:** This kit is available for $44.99 and includes a carrying case, extra strings, a strap, a pick set, and a pitch pipe.\n\n**Summary:** All three options are affordable and suitable for beginners, with prices ranging from $30.99 to $49.99.\n\n**Metadata:**\n*   **Smartxchoices:** Color: Blue, Rating: 4.2, R

<p>Q: "Which one is the best option?"</p>

<p>A: Here's a comparison of the three guitar options based on the provided information:</p>
<p><strong>Smartxchoices 38" Acoustic Guitar Bundle Kit:</strong></p>
<ul>
<li><strong>Price:</strong> $30.99</li>
<li><strong>Color:</strong> Blue</li>
<li><strong>Rating:</strong> 4.2 (2372 ratings)</li>
<li><strong>Features:</strong> Includes a gig bag, extra steel strings, strap, pitch pipe, and pick.</li>
<li><strong>Summary:</strong> This is the most affordable option and comes with a good selection of accessories.</li>
</ul>
<p><strong>Best Choice Products 30in Kids Acoustic Guitar:</strong></p>
<ul>
<li><strong>Price:</strong> $44.99</li>
<li><strong>Color:</strong> SoCal Green</li>
<li><strong>Rating:</strong> 4.3 (5251 ratings)</li>
<li><strong>Features:</strong> Includes a carrying case, extra strings, strap, pick set, and pitch pipe. It has nylon strings, which are easier for beginners.</li>
<li><strong>Summary:</strong> A slightly more expensive option with a carrying case and nylon strings, making it suitable for younger beginners.</li>
</ul>
<p><strong>Ashthorpe 38-inch Beginner Acoustic Guitar Package:</strong></p>
<ul>
<li><strong>Price:</strong> $49.99</li>
<li><strong>Color:</strong> Blue</li>
<li><strong>Rating:</strong> 4.4 (4608 ratings)</li>
<li><strong>Features:</strong> Includes a gig bag, strings, strap, tuner, pitch pipe, and instructional sheet. It's made with laminated basswood.</li>
<li><strong>Summary:</strong> The most expensive option, but it includes a tuner and instructional sheet, and is made with basswood.</li>
</ul>
<p><strong>Comparison:</strong></p>
<p>The <strong>Ashthorpe</strong> package has the highest rating (4.4) and includes a tuner and instructional sheet. However, it's also the most expensive at $49.99. The <strong>Best Choice Products</strong> guitar is a good middle ground at $44.99, with nylon strings and a carrying case. The <strong>Smartxchoices</strong> kit is the most budget-friendly at $30.99, offering a good set of accessories.</p>

In [6]:
print_chat_history_db()

📁 Chat History DB Path: /Users/muhammetsantas/Downloads/EDUCATION/POWERCODERS/PERSONAL_PROJECTS/chatbot/project-25-2-scrum-team-data/chat_history.db
✅ Is there DB? True

📊 Tables: [('chat_sessions',), ('sqlite_sequence',)]

📝 Total sessions: 1


🔹 Session 1:
   ID: 1
   Session ID: 609387
   Created At: 2025-12-11 18:19:48
   Updated At: 2025-12-11 18:20:16
   Messages (4 entries):
      1. Role: user | I want to buy a cheap guitar for beginners. What are the most cheapest 3 options available?
      2. Role: assistant | Here are the three cheapest guitar options for beginners based on the provided information:

1.  **Smartxchoices starter kit:** Priced at $30.99, this kit includes everything you need to get started, ...
      3. Role: user | Which one is the best option?
      4. Role: assistant | Here's a comparison of the three guitar options to help determine the best choice:

**Smartxchoices 38" Kids Acoustic Guitar Bundle Kit:**

*   **Price:** $30.99
*   **Color:** Blue
*   **Rat

{'609387': {'messages': [{'role': 'user',
    'content': 'I want to buy a cheap guitar for beginners. What are the most cheapest 3 options available?'},
   {'role': 'assistant',
    'content': 'Here are the three cheapest guitar options for beginners based on the provided information:\n\n1.  **Smartxchoices starter kit:** Priced at $30.99, this kit includes everything you need to get started, such as a guitar, strings, a pitch pipe, a shoulder strap, and a carrying case.\n2.  **Best Choice Products 38-inch Beginner Acoustic Guitar Package:** This option costs $49.99 and comes with a gig bag, picks, a strap, a tuner, and an instructional sheet.\n3.  **Best Choice Products 30in Kids Acoustic Guitar Starter Kit:** This kit is available for $44.99 and includes a carrying case, extra strings, a strap, a pick set, and a pitch pipe.\n\n**Summary:** All three options are affordable and suitable for beginners, with prices ranging from $30.99 to $49.99.\n\n**Metadata:**\n*   **Smartxchoices:** C

In [7]:
%%time
query = "I have 32 dollars. Which one can I buy with that budget?"

response = ask_question(question=query, k=5, use_reranker=False, session_id=session_id)

print("Question:", query)
print("\nAnswer:\n", response)

history: [{'role': 'user', 'content': 'I want to buy a cheap guitar for beginners. What are the most cheapest 3 options available?'}, {'role': 'assistant', 'content': 'Here are the three cheapest guitar options for beginners based on the provided information:\n\n1.  **Smartxchoices starter kit:** Priced at $30.99, this kit includes everything you need to get started, such as a guitar, strings, a pitch pipe, a shoulder strap, and a carrying case.\n2.  **Best Choice Products 38-inch Beginner Acoustic Guitar Package:** This option costs $49.99 and comes with a gig bag, picks, a strap, a tuner, and an instructional sheet.\n3.  **Best Choice Products 30in Kids Acoustic Guitar Starter Kit:** This kit is available for $44.99 and includes a carrying case, extra strings, a strap, a pick set, and a pitch pipe.\n\n**Summary:** All three options are affordable and suitable for beginners, with prices ranging from $30.99 to $49.99.\n\n**Metadata:**\n*   **Smartxchoices:** Color: Blue, Rating: 4.2, R

Question: I have 32 dollars. Which one can I buy with that budget?

Answer:
 <p>You can purchase the Smartxchoices starter kit with a budget of $32. It is priced at $30.99.</p>
<p>Here's a breakdown of the options:</p>
<ul>
<li><strong>Smartxchoices starter kit:</strong> $30.99</li>
<li><strong>Best Choice Products 38-inch Beginner Acoustic Guitar Package:</strong> $49.99</li>
<li><strong>Best Choice Products 30in Kids Acoustic Guitar Starter Kit:</strong> $44.99</li>
</ul>
CPU times: user 772 ms, sys: 83.7 ms, total: 855 ms
Wall time: 12.8 s
Question: I have 32 dollars. Which one can I buy with that budget?

Answer:
 <p>You can purchase the Smartxchoices starter kit with a budget of $32. It is priced at $30.99.</p>
<p>Here's a breakdown of the options:</p>
<ul>
<li><strong>Smartxchoices starter kit:</strong> $30.99</li>
<li><strong>Best Choice Products 38-inch Beginner Acoustic Guitar Package:</strong> $49.99</li>
<li><strong>Best Choice Products 30in Kids Acoustic Guitar Starter Kit:</strong> $44.99</li>
</ul>

In [8]:
print_chat_history_db()

📁 Chat History DB Path: /Users/muhammetsantas/Downloads/EDUCATION/POWERCODERS/PERSONAL_PROJECTS/chatbot/project-25-2-scrum-team-data/chat_history.db
✅ Is there DB? True

📊 Tables: [('chat_sessions',), ('sqlite_sequence',)]

📝 Total sessions: 1


🔹 Session 1:
   ID: 1
   Session ID: 609387
   Created At: 2025-12-11 18:19:48
   Updated At: 2025-12-11 18:21:01
   Messages (6 entries):
      1. Role: user | I want to buy a cheap guitar for beginners. What are the most cheapest 3 options available?
      2. Role: assistant | Here are the three cheapest guitar options for beginners based on the provided information:

1.  **Smartxchoices starter kit:** Priced at $30.99, this kit includes everything you need to get started, ...
      3. Role: user | Which one is the best option?
      4. Role: assistant | Here's a comparison of the three guitar options to help determine the best choice:

**Smartxchoices 38" Kids Acoustic Guitar Bundle Kit:**

*   **Price:** $30.99
*   **Color:** Blue
*   **Rat

{'609387': {'messages': [{'role': 'user',
    'content': 'I want to buy a cheap guitar for beginners. What are the most cheapest 3 options available?'},
   {'role': 'assistant',
    'content': 'Here are the three cheapest guitar options for beginners based on the provided information:\n\n1.  **Smartxchoices starter kit:** Priced at $30.99, this kit includes everything you need to get started, such as a guitar, strings, a pitch pipe, a shoulder strap, and a carrying case.\n2.  **Best Choice Products 38-inch Beginner Acoustic Guitar Package:** This option costs $49.99 and comes with a gig bag, picks, a strap, a tuner, and an instructional sheet.\n3.  **Best Choice Products 30in Kids Acoustic Guitar Starter Kit:** This kit is available for $44.99 and includes a carrying case, extra strings, a strap, a pick set, and a pitch pipe.\n\n**Summary:** All three options are affordable and suitable for beginners, with prices ranging from $30.99 to $49.99.\n\n**Metadata:**\n*   **Smartxchoices:** C

In [14]:
%%time
query = "OK, I want to buy Smartxchoices, is there any red color of this guitar?"

response = ask_question(question=query, k=5, use_reranker=False, session_id=session_id)

print("Question:", query)
print("\nAnswer:\n", response)

✅ API key was loaded successfully!
📍 .env file location: /Users/muhammetsantas/Downloads/EDUCATION/POWERCODERS/PERSONAL_PROJECTS/chatbot/project-25-2-scrum-team-data/.env
LLM instance created successfully.
history: [{'role': 'user', 'content': 'I want to buy a cheap guitar for beginners. What are the most cheapest 3 options available?'}, {'role': 'assistant', 'content': 'Here are the three cheapest guitar options for beginners based on the provided information:\n\n1.  **Smartxchoices starter kit (acoustic):** Priced at $30.99, this kit includes everything you need to get started, such as a carrying case, guitar pick, pitch pipe, shoulder strap, and an extra set of strings. It has a rating of 4.2 stars from 2372 ratings.\n2.  **Best Choice Products 30in Kids Acoustic Guitar:** This option is available for $44.99 and includes a carrying case, extra strings, strap, pick set, and pitch pipe. It has a rating of 4.3 stars.\n3.  **Ashthorpe 38-inch Beginner Acoustic Guitar Package:** This pac

Question: OK, I want to buy Smartxchoices, is there any red color of this guitar?

Answer:
 <p>The provided product information does not include this detail. However, the Smartxchoices 38" Kids Acoustic Guitar Bundle Kit is available in Blue and is priced at $30.99 with a rating of 4.2 based on 2372 ratings.</p>
<p><strong>Key Features</strong></p>
<ul>
<li>Designed for beginner music lovers</li>
<li>Includes gig bag, extra set steel strings, strap, pitch pipe, and pick</li>
</ul>
<p><strong>Specifications / Metadata</strong></p>
<ul>
<li>Color: Blue</li>
<li>Store: Smartxchoices</li>
<li>Rating: 4.2</li>
<li>Rating Count: 2372</li>
<li>Price: $30.99</li>
</ul>

In [15]:
print_chat_history_db()

📁 Chat History DB Path: /Users/muhammetsantas/Downloads/EDUCATION/POWERCODERS/PERSONAL_PROJECTS/chatbot/project-25-2-scrum-team-data/chat_history.db
✅ DB var mı? True

📊 Tablolar: [('chat_sessions',), ('sqlite_sequence',)]

📝 Toplam session sayısı: 2


🔹 Session 1:
   ID: 2
   Session ID: 315921
   Created At: 2025-12-10 21:20:34
   Updated At: 2025-12-10 21:20:34
   Messages (4 entry):
      1. Role: user | Content: I want to buy a cheap guitar for beginners. What are the most cheapest 3 options available?...
      2. Role: assistant | Content: Here are the three cheapest guitar options for beginners based on the provided information:

1.  **S...
      3. Role: user | Content: Which one is the best option?...
      4. Role: assistant | Content: Here's a comparison of the three guitar options to help determine the "best" choice, based on the pr...

🔹 Session 2:
   ID: 6
   Session ID: 780887
   Created At: 2025-12-10 21:48:54
   Updated At: 2025-12-10 21:48:54
   Messages (8 entry):
  

In [9]:
import sqlite3
import pandas as pd

# Connect to the SQLite database
conn = sqlite3.connect(str(db_path))

# Read the chat_sessions table into a pandas DataFrame
df_chat_history = pd.read_sql_query("SELECT * FROM chat_sessions", conn)

# Close the connection
conn.close()

# Display the DataFrame
df_chat_history

id             session_id  \
0   1                 609387   
1   2  4j0vJopmtCt9gI8OBl4Vg   
2   3           TEST_SESSION   
3   4      FRONTEND_TEST_123   
4   5        default_session   
5   6  smo5x50CGRuN6Q3odgKfG   
6   7  SNaQD_ERCnFOncSptjnzS   
7   8  awPCKdp_-Aj-d0lpSm6EZ   
8   9  6xInlwu5Zng5tEUusjtLR   

                                            messages  \
0  [{"role": "user", "content": "I want to buy a ...   
1  [{"role": "user", "content": "hello"}, {"role"...   
2  [{"role": "user", "content": "What is the chea...   
3  [{"role": "user", "content": "What are the top...   
4  [{"role": "user", "content": "hello"}, {"role"...   
5  [{"role": "user", "content": "hi"}, {"role": "...   
6  [{"role": "user", "content": "hi"}, {"role": "...   
7  [{"role": "user", "content": "hi"}, {"role": "...   
8  [{"role": "user", "content": "hi"}, {"role": "...   

                                        html_answers           created_at  \
0  ["<p>Here are the three cheapest guitar option...  2025-12-11 18:19:48   
1        ["<p>Hello! How can I help you today?</p>"]  2025-12-11 18:31:59   
2  ["<p>The cheapest guitar is the Smartxchoices ...  2025-12-11 18:42:10   
3  ["<p>Here are three affordable guitar options ...  2025-12-11 18:43:41   
4  ["<p>Hello! How can I help you today?</p>", "<...  2025-12-11 18:46:38   
5  ["<p>Hi there! How can I help you today?</p>",...  2025-12-11 19:23:05   
6  ["<p>Hello! How can I help you today?</p>", "<...  2025-12-11 19:29:02   
7  ["<p>Hi there! How can I help you today?</p>",...  2025-12-11 19:34:14   
8  ["<p>Hello! How can I help you today?</p>", "<...  2025-12-11 19:53:13   

            updated_at  
0  2025-12-11 18:21:01  
1  2025-12-11 18:31:59  
2  2025-12-11 18:42:10  
3  2025-12-11 18:43:41  
4  2025-12-11 19:18:26  
5  2025-12-11 19:25:56  
6  2025-12-11 19:32:34  
7  2025-12-11 19:39:33  
8  2025-12-11 19:54:49